code/src/context/I18nProvider.tsx

```diff
@@ -1,18 +1,66 @@
-import { FC, ReactNode } from "react";
+import {
+  Dispatch,
+  PropsWithChildren,
+  SetStateAction,
+  createContext,
+  useContext,
+  useState
+} from "react";
 import { IntlProvider } from "react-intl";
 
+import { LOCAL_STORAGE_KEYS } from "@/constants/common";
 import messages from "@/messages";
 
-type I18nProividerProps = {
-  children: ReactNode;
+type I18nProviderProps = Required<PropsWithChildren>;
+
+export type Locale = "en" | "uk";
+
+const SUPPORTED_LOCALES: Locale[] = ["en", "uk"];
+
+type LocaleContextType = {
+  locale: Locale;
+  setLocale: Dispatch<SetStateAction<Locale>>;
+};
+
+const LocaleContext = createContext<LocaleContextType | null>(null);
+
+const getInitialLocale = (): Locale => {
+  const storedLocale = window.localStorage.getItem(LOCAL_STORAGE_KEYS.locale);
+  const browserLocale = window.navigator.language.split("-")[0] as Locale;
+
+  if (SUPPORTED_LOCALES.includes(storedLocale as Locale)) {
+    return storedLocale as Locale;
+  }
+  if (SUPPORTED_LOCALES.includes(browserLocale)) {
+    return browserLocale;
+  }
+  return "en";
 };
```

Comment:
I suggest improving this logic like follows:

```tsx
const getInitialLocale = (): Locale => {
  const storedLocale = window.localStorage.getItem(
    LOCAL_STORAGE_KEYS.locale
  ) as Locale;

  if (SUPPORTED_LOCALES.includes(storedLocale)) {
    return storedLocale;
  }

  const browserLocale = window.navigator.language.split("-")[0] as Locale;

  if (SUPPORTED_LOCALES.includes(browserLocale)) {
    return browserLocale;
  }
  
  return "en";
};
```

This way we can exit early when `storedLocale` is provided and maybe save a little performance. Also in my example I moved `as` casting to the variables to make checks more clean. What do you think about my example?

Reply:
Thank you. It's looking pretty nice